[STORM-2704] Check blob permission before submitting the topology

Ethanlm opened the pull request:
See: https://issues.apache.org/jira/browse/STORM-2704

The original code only checks if the keys in `topology.blobstore.map` exist in the BlobStore.  We need to check the ACL as well.

Ethanlm commented:
Tested it manually. All the experiments submitted topologies as `ethan`.

1. With `storm.blobstore.acl.validation.enabled: false`
1.1. When the key does not exist: `InvalidTopologyException` 
![image](https://user-images.githubusercontent.com/14900612/29582138-2ec1c412-8741-11e7-9e0f-b080acb908ff.png)
1.2. When the key exists with ACL set to `u:mapredqa:rwa`: submitted successfully.
![image](https://user-images.githubusercontent.com/14900612/29582141-31f2dbee-8741-11e7-9aa5-faa688132b98.png)

2. With `storm.blobstore.acl.validation.enabled: true`
2.1 When the key does not exist: `InvalidTopologyException` 
![image](https://user-images.githubusercontent.com/14900612/29582188-62d0b7ae-8741-11e7-9621-a52bed416a5e.png)
2.2 When the key exists with ACL set to `u:mapredqa:rwa`: `AuthorizationException`
![image](https://user-images.githubusercontent.com/14900612/29582211-76295108-8741-11e7-8aa4-b1c2dc8294d1.png)
2.3 When the key exists with ACL set to `u:ethan:rwa`: submitted successfully.
![image](https://user-images.githubusercontent.com/14900612/29582236-9607c590-8741-11e7-9f24-36ea0a590c7b.png)

HeartSaVioR commented on storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java:
```diff
@@ -2529,7 +2529,7 @@ public void submitTopologyWithOpts(String topoName, String uploadedJarLocation,
             }
             validateTopologyWorkerMaxHeapSizeConfigs(topoConf, topology);
-            Utils.validateTopologyBlobStoreMap(topoConf, Sets.newHashSet(blobStore.listKeys()));
+            Utils.validateTopologyBlobStoreMap(topoConf);
```
Maybe better to add method which contains NimbusBlobStore as parameter to avoid creating NimbusBlobStore in Nimbus.

Ethanlm replied:
Will see how to do it

HeartSaVioR commented on storm-client/test/jvm/org/apache/storm/TestConfigValidate.java:
```diff
@@ -84,21 +83,6 @@ public void invalidConfigTest() throws InvocationTargetException, NoSuchMethodEx
         ConfigValidation.validateFields(conf);
     }
 
-    @Test(expected = InvalidTopologyException.class)
```
Could we mock NimbusBlobStore and retain the test?

Ethanlm replied:
I tried to mock it and then decided not to because the logic is really simple in `validateTopologyBlobStoreMap`. But I will add the unit test if you think it's better

HeartSaVioR replied:
Yes please do it if you really don't mind. I think we need to have broader coverage of tests.

Ethanlm replied:
Sure. Will do

Ethanlm commented:
@HeartSaVioR  Thanks for the review. Addressed your comments. 

Tested it manually again.

Also ran the same tests with commenting out [validateTopologyBlobStoreMap function in validateConfs](https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/StormSubmitter.java#L564-L567) so that I can test the function call  (`Utils.validateTopologyBlobStoreMap(topoConf, blobStore)` ) in `Nimbus.java`. All tests are passed.  Some exception messages like:

![image](https://user-images.githubusercontent.com/14900612/29624442-b3987ade-87ee-11e7-9143-49ba056d991c.png)
![image](https://user-images.githubusercontent.com/14900612/29624448-b61529ec-87ee-11e7-98ec-3411e386560d.png)
